File: Permutations.java
public class Solution {
    private void divideAndConquer(List<Integer> selects, List<Integer> num, List<List<Integer>> result) {
        if (num.isEmpty()) {
            result.add(new ArrayList<Integer>(selects));
            return;
        }
        for (int i = 0; i < num.size(); i++) {
            int n = num.get(i);
            num.remove(i);
            selects.add(n);
            divideAndConquer(selects, num, result);
            selects.remove(selects.size() - 1);
            num.add(i, n);
        }
    }
    public List<List<Integer>> permute(int[] num) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> selects = new ArrayList<Integer>();
        List<Integer> numbers = new ArrayList<Integer>();
        for (int i = 0; i < num.length; i++)
            numbers.add(num[i]);
        divideAndConquer(selects, numbers, result);
        return result;
    }
}

File: Permutations.cpp
class Solution {
public:
    void divideAndConquer(vector<int>& selects, vector<int>& num, vector<vector<int> >& result) {
        if (num.empty()) {
            result.push_back(selects);
            return;
        }
        for (int i = 0; i < num.size(); i++) {
            selects.push_back(num[i]);
            int temp = num[i];
            num[i] = num.back();
            num.pop_back();
            divideAndConquer(selects, num, result);
            num.push_back(num[i]);
            num[i] = temp;
            selects.pop_back();
        }
    }
    vector<vector<int> > permute(vector<int> &num) {
        vector<vector<int> > result;
        vector<int> selects;
        divideAndConquer(selects, num, result);
        return result;
    }
};